RFM9x.device.nut

```diff
@@ -0,0 +1,404 @@
+const RFM9X_REG_SYMB_TIMEOUT_LSB = 0x1F;
+
+const RFM9X_REG_PREAMBLE_MSB = 0x20;
+const RFM9X_REG_PREAMBLE_LSB = 0x21;
+const RFM9X_REG_PAYLOAD_LENGTH = 0x22;
+const RFM9X_REG_MAX_PAYLOAD_LENGTH = 0x23;
+const RFM9X_REG_HOP_PERIOD = 0x24;
+const RFM9X_REG_FIFO_RX_BYTE_ADDR = 0x25;
+
+
+// -------------------- Shared Registers ---------------------- //
+
+// DIO mappings are used for interrupts
+
+const RFM9X_REG_DIO_MAPPING1 = 0x40;
+const RFM9X_REG_DIO_MAPPING2 = 0x41;
+const RFM9X_REG_VERSION = 0x42;
+const RFM9X_REG_TCXO = 0x4B;
+
+// Default init constants
+const RFM9X_DEFAULT_FREQUENCY = 915000000;
+const RFM9X_DEFAULT_BANDWIDTH = "125";
+const RFM9X_DEFAULT_SF = 7;
+const RFM9X_DEFAULT_CODING_RATE = "4/5";
+const RFM9X_DEFAULT_IMPLICIT_HEADER_MODE = 0;
+const RFM9X_DEFAULT_PREAMBLE_LENGTH = 8;
+
+const RFM9X_FREQ_STEP = 61.035;
+
+// LoRa Mode
+const RFM9X_CONFIG_BYTE = 0x80; 
+
+// Interrupt Masks AND flags
+enum RFM9X_FLAGS {
+    CAD_DETECTED,
+    FHSS_CHANGE_CHANNEL,
+    CAD_DONE,
+    TX_DONE,
+    VALID_HEADER,
+    PAYLOAD_CRC_ERROR,
+    RX_DONE,
+    RX_TIMEOUT
+};
+
+// Operating modes
+const RFM9X_SLEEP = 0x00;
+const RFM9X_STANDBY = 0x01;
+const RFM9X_FSTX = 0x02;
+const RFM9X_TX = 0x03;
+const RFM9X_FSRX = 0x04;
+const RFM9X_RXCONTINUOUS = 0x05;
+const RFM9X_RXSINGLE = 0x06;
+const RFM9X_CAD = 0x07;
+
+
+class RFM9x {
+    static VERSION = "0.1.0";
+
+    // -------------------- Some Tables --------------------------- //
+    
+    // This table maps bandwidth strings (in kHz) to the values which go in the 
+    // corresponding register
+    static BWTABLE = {
+        "7.8" : 0x00,
+        "10.4" : 0x01,
+        "15.6" : 0x02,
+        "20.8" : 0x03,
+        "31.25": 0x04,
+        "41.7": 0x05,
+        "62.5": 0x06,
+        "125": 0x07,
+        "250" : 0x08,
+        "500": 0x09
+    };
+    
+    // This table maps coding rate strings to the values which go in the corresponding
+    // register
+    static CRTABLE = {
+        "4/5": 0x01,
+        "4/6": 0x02,
+        "4/7": 0x03,
+        "4/8": 0x04
+    };
+
+    _spiModule = null;
+    _cs = null;
+    _irqPin = null;
+    _receiveHandler = null;
+    _sendFinished = true;
+    _error = false;    
+    
+    constructor(spi, intPin, cs=null) {
+        // Assume spi is already initialized, assume cs is already initialized
+        _spiModule = spi;
+        _irqPin = intPin;
+        _cs = cs;
+    }
+    
+    function init() {
+        setMode(RFM9X_SLEEP); // need to do this to go into LoRa mode
+        
+        // Defaults
+        setFrequency(RFM9X_DEFAULT_FREQUENCY);
+        setBandwidth(RFM9X_DEFAULT_BANDWIDTH);
+        setSpreadingFactor(RFM9X_DEFAULT_SF);
+        setCodingRate(RFM9X_DEFAULT_CODING_RATE);
+        setImplicitHeaderMode(RFM9X_DEFAULT_IMPLICIT_HEADER_MODE);
+        setPreambleLength(RFM9X_DEFAULT_PREAMBLE_LENGTH);
+
+        _configISR();
+        
+        _setInLoRaMode();
+
+    }
+    
+    function sendData(data) {
```

Comment:
It looks like the module doesn't allow you to send another packet until the previous one succeeds (please correct me if this assumption is wrong). So it probably makes sense to have per library callback `onSent` or anything.